sources/ghostdog/src/nvme.rs

```diff
@@ -0,0 +1,362 @@
+            maxdna: [__u8::default(); 16],
+            maxcna: __le32::default(),
+            oaqd: __le32::default(),
+            rsvd568: [__u8::default(); 200],
+            subnqn: [__char::default(); 256],
+            rsvd1024: [__u8::default(); 768],
+            ioccsz: __le32::default(),
+            iorcsz: __le32::default(),
+            icdoff: __le16::default(),
+            fcatt: __u8::default(),
+            msdbd: __u8::default(),
+            ofcs: __le16::default(),
+            dctype: __u8::default(),
+            rsvd1807: [__u8::default(); 241],
+            psd: [NvmeIdPsd::default(); 32],
+            vs: [__u8::default(); 1024],
+        }
+    }
+}
+
+#[repr(C)]
+#[derive(Default, Debug)]
+/// Represents an NVME command send through ioctl
+struct NvmePassthruCmd {
+    opcode: __u8,
+    flags: __u8,
+    rsvd1: __u16,
+    nsid: __u32,
+    cdw2: __u32,
+    cdw3: __u32,
+    metadata: __u64,
+    addr: __u64,
+    metadata_len: __u32,
+    data_len: __u32,
+    cdw10: __u32,
+    cdw11: __u32,
+    cdw12: __u32,
+    cdw13: __u32,
+    cdw14: __u32,
+    cdw15: __u32,
+    timeout_ms: __u32,
+    result: __u32,
+}
+
+#[derive(Debug)]
+/// Represents an EBS NVME volume
+pub(crate) struct EBSNvmeDevice {
+    pub(crate) device_name: String,
+}
+
+impl From<NvmeIdCtrl> for EBSNvmeDevice {
+    fn from(value: NvmeIdCtrl) -> Self {
+        // The device name is within the first 32 bytes of the vendor specific data, which
+        // is guaranteed to be 1024 bytes long. Trim the resulting string, otherwise udev
+        // doesn't create the correct symlinks.
+        let device_name = String::from_utf8_lossy(&value.vs[0..32]).trim().to_string();
+
+        Self { device_name }
+    }
+}
+
+// Define the nvme_admin_command ioctl function; use `readwrite` because the syscall
+// is defined IOWR.
+// See: https://elixir.bootlin.com/linux/v6.1/source/include/uapi/linux/nvme_ioctl.h#L98
+ioctl_readwrite!(
+    nvme_admin_command,
+    NVME_IOC_MAGIC,
+    NVME_IOC_ADMIN_CMD,
+    NvmePassthruCmd
+);
+
+/// Retrieves the device information for the passed EBS volume
+pub(crate) fn query_ebs_nvme_device(path: &PathBuf) -> Result<EBSNvmeDevice> {
+    let device = fs::File::open(path)
+        .context(error::DeviceOpenSnafu { path })?
+        .into_raw_fd();
+
+    // The structure used to recieve the result of the ioctl syscall, a pointer
+    // to this struct will be send in the ioctl command.
+    let mut ctrl = NvmeIdCtrl::default();
+    let mut cmd = NvmePassthruCmd {
+        opcode: NVME_ADMIN_IDENTIFY,
+        data_len: std::mem::size_of::<NvmeIdCtrl>() as u32,
+        addr: addr_of_mut!(ctrl) as u64,
+        // This is required, otherwise the syscall always returns the same device; this
+        // value corresponds to what Amazon Linux does.
+        // See: https://github.com/amazonlinux/amazon-ec2-utils/blob/09b53f2a20c3e4d227acfb452138fdb672baaa1a/ebsnvme-id#L114
+        cdw10: 1,
```

Comment:
This is defined in [NVME 1.1](https://www.nvmexpress.org/wp-content/uploads/NVM-Express-1_1a.pdf), Figure 81 (page 84). The value 1 means "The Identify Controller data structure is returned to
the host."